pybamm/solvers/rk4_ode_solver.py

```diff
@@ -0,0 +1,186 @@
+            if not model.events:
+                new_event_signs = np.array([1, 1])
+            else:
+                new_event_signs = np.sign(
+                    np.concatenate(
+                        [
+                            event(t_eval[k + 1], xk.full().flatten(), inputs)
+                            for event in model.terminate_events_eval
+                        ]
+                    )
+                )
+            # Exit loop if the sign of an event changes
+            # Locate the event time using a root finding algorithm and
+            # event state using interpolation. The solution is then truncated
+            # so that only the times up to the event are returned
+            i = 1
+            if (new_event_signs != init_event_signs).any():
+                t_event = t_eval[k]
+                y_event = y_sol[:, k]
+                pre_event_signs = init_event_signs
+                s = 1
+                dt = dt / 2
+                while (i < self.max_step_decrease_count) and (dt > 1e-6):
+                    i += 1
+                    xk = F(t0=t_event, y0=y_event, dt=dt * s)["yf"]
+                    t_event = t_event + s * dt
+                    y_event = xk.full().flatten()
+
+                    new_event_signs = np.sign(
+                        np.concatenate(
+                            [
+                                event(t_event, y_event, inputs)
+                                for event in model.terminate_events_eval
+                            ]
+                        )
+                    )
+                    if (new_event_signs != pre_event_signs).any():
+                        s = -1 * s
+                    dt = dt / 2
+                    pre_event_signs = new_event_signs
+                y_sol[:, k + 1] = y_event
+                break
+
+        if i > 1:
+            # return truncated solution
+            t_truncated = np.append(t_eval[: k + 1], t_event)
+            y_trunctaed = y_sol[:, : k + 2].full()
+            truncated_step_sol = pybamm.Solution(t_truncated, y_trunctaed)
+            # assign temporary solve time
+            truncated_step_sol.solve_time = np.nan
+            # append solution from the current step to solution
+            solution = truncated_step_sol
+
+            solution.termination = "event"
+            solution.t_event = t_event
+            solution.y_event = y_event
+        else:
+            # return full solution
+            step_sol = pybamm.Solution(t_eval, y_sol.full())
+            # assign temporary solve time
+            step_sol.solve_time = np.nan
+            # append solution from the current step to solution
+            solution = step_sol
+
+            solution.termination = "final"
+
+        return solution, F
+
+    def _run_integrator(self, integrator, model, y0, inputs, t_eval):
+        try:
+            sol = integrator
+            y_values = sol.y
+            return pybamm.Solution(sol.t, y_values)
```

Comment:
I don't thing the `get_integrator` / `run_integrator` structure is appropriate in this case. Here, the integrator is not a function, but already the solution, and `get_integrator` and `run_integrator` are only called once each. So you could put the whole thing in the `_integrate` function